src/lm/diff.cpp

```diff
@@ -0,0 +1,255 @@
+/**
+ * @file diff.cpp
+ * @author Sean Massung
+ */
+
+#include <algorithm>
+#include <queue>
+#include "lm/diff.h"
+#include "porter2_stemmer.h"
+
+namespace meta
+{
+namespace lm
+{
+diff::diff(const cpptoml::table& config) : lm_{config}
+{
+    auto table = config.get_table("diff");
+    if (!table)
+        throw diff_exception{"missing [diff] table from config"};
+
+    auto nval = table->get_as<int64_t>("n-value");
+    if (!nval)
+        throw diff_exception{"n-value not specified in config"};
+    n_val_ = *nval;
+
+    auto edits = table->get_as<int64_t>("max-edits");
+    if (!edits)
+        throw diff_exception{"max-edits not specified in config"};
+    max_edits_ = *edits;
+
+    auto b_pen = table->get_as<double>("base-penalty");
+    base_penalty_ = b_pen ? *b_pen : 0.0;
+
+    auto i_pen = table->get_as<double>("insert-penalty");
+    insert_penalty_ = i_pen ? *i_pen : 0.0;
+
+    auto s_pen = table->get_as<double>("substitute-penalty");
+    substitute_penalty_ = s_pen ? *s_pen : 0.0;
+
+    auto r_pen = table->get_as<double>("remove-penalty");
+    remove_penalty_ = r_pen ? *r_pen : 0.0;
+
+    set_stems(*table);
+    set_function_words(*table);
+}
+
+std::vector<std::pair<sentence, double>>
+    diff::candidates(const sentence& sent, bool use_lm /* = false */)
+{
+    use_lm_ = use_lm;
+    using pair_t = std::pair<sentence, double>;
+    auto comp = [](const pair_t& a, const pair_t& b)
+    {
+        return a.second < b.second;
+    };
+    std::priority_queue<pair_t, std::vector<pair_t>, decltype(comp)> candidates{
+        comp};
+    add(candidates, sent);
+
+    seen_.clear();
+    step(sent, candidates, 0);
+
+    std::vector<pair_t> sorted;
```

Comment:
Candidates should be small enough that it doesn't matter (e.g. 20), but I can add it anyway.